### PR TITLE
networkd: support drop-in directories for .network files (#3655)

### PR DESCRIPTION
**Pull request by Sufflope**

### CODE REVIEW

**Comment by martinpitt** [comment]
This breaks the networkd test all over, so please don't land this.

**Comment by Sufflope** [comment]
> This breaks the networkd test all over

Yeah I was able to notice that.

I've been working on it on my spare time, and QEMU-based tests don't work on my Fedora 24 station for whatever reason (VM hangs after boot, doing nothing and never exiting).

I'm trying to get more information by using an Ubuntu VM to be more like the integration tests platform but slowly progressing.

Anyway I don't think anyone was going to merge a PR which doesn't build, right?

> so please don't land this.

As-is, indeed ; or should I completely stop trying to contribute?

**Comment by martinpitt** [comment]
Note that you can also run the tests in an nspawn (or LXC) container, or even on your local machine.

> As-is, indeed ; or should I completely stop trying to contribute?

Urgh, "as-is" of course -- I was just saying that we cannot land this because it causes a regression. Of course that did not mean "please stop working", we do value (and get a lot of) contributions! Thanks!

File: src/network/networkd-network.c [hunk 1]
```diff
@@ -40,6 +40,10 @@ static int network_load_one(Manager *manager, const char *filename) {
         _cleanup_network_free_ Network *network = NULL;
         _cleanup_fclose_ FILE *file = NULL;
         char *d;
+        _cleanup_free_ const char *dropin_dirname = NULL;
+        _cleanup_strv_free_erase_ char **dropin_dirs = NULL;
```
Comment (poettering) [inline src/network/networkd-network.c]: why "erase"? that's only used for stuff that like passphrases and cryptographic keys...

File: src/network/networkd-network.c [hunk 2]
```diff
@@ -40,6 +40,10 @@ static int network_load_one(Manager *manager, const char *filename) {
         _cleanup_network_free_ Network *network = NULL;
         _cleanup_fclose_ FILE *file = NULL;
         char *d;
+        _cleanup_free_ const char *dropin_dirname = NULL;
```
Comment (poettering) [inline src/network/networkd-network.c]: hmm, does this compile without warnings? strings that are allocated dynamically cannot be const in C.

File: src/network/networkd-network.c [hunk 3]
```diff
@@ -136,7 +140,17 @@ static int network_load_one(Manager *manager, const char *filename) {
         network->proxy_arp = -1;
         network->ipv6_accept_ra_use_dns = true;
 
-        r = config_parse(NULL, filename, file,
+        dropin_dirname = strjoin("/", network->name, ".network.d", NULL);
```
Comment (poettering) [inline src/network/networkd-network.c]: you could just use strjoina() here (i.e. use stack allocation), after all the name is pretty much bounded. You can then drop _cleanup_free_ from the variable declaration, but keep the "const" for it...

File: src/network/networkd-network.c [hunk 4]
```diff
@@ -136,7 +140,17 @@ static int network_load_one(Manager *manager, const char *filename) {
...
+
+        r = strv_extend_strv_concat(&dropin_dirs, (char**) network_dirs, dropin_dirname);
+        if (r < 0)
+            return r;
```
Comment (poettering) [inline src/network/networkd-network.c]: 8ch indentation please, see CODING_STYLE...

File: src/network/networkd-network.c [hunk 5]
```diff
@@ -136,7 +140,17 @@ static int network_load_one(Manager *manager, const char *filename) {
...
+
+        r = strv_make_nulstr(dropin_dirs, &dropin_dirs_nulstr, &dropin_dirs_nulstr_size);
+        if (r < 0)
+            return r;
```
Comment (poettering) [inline src/network/networkd-network.c]: similar here...

**Comment by poettering** [comment]
patch looks pretty OK... found some minor issues only.

**Comment by Sufflope** [comment]
> Note that you can also run the tests in an nspawn (or LXC) container, or even on your local machine.

At first I tried to run them in an nspawn container (I'm a bit relunctant to run some code as root on my real machine, especially when I hacked it :smile: ) but the integration tests failed because trouble with loop devices. After trying everything I found about  adding capabilities and DeviceAllow to the container I gave up. I must have missed somthing.

> I was just saying that we cannot land this because it causes a regression. Of course that did not mean "please stop working", we do value (and get a lot of) contributions! Thanks!

Sorry, I overreacted. I'll rework it!

**Comment by keszybz** [comment]
Looks good, but needs a man page update.

**Comment by keszybz** [comment]
Man page update in the commit shown above.

I think we need to do the same for links, though.

**Comment by poettering** [comment]
patch looks ok, but the new test fails in the CI, @Sufflope please have a look!

**Comment by martinpitt** [comment]
Superseded by PR #4123

**Pull request closed without merging**